c10/cuda/CUDAStream.cpp

```diff
@@ -71,6 +70,11 @@ static std::array<LeakyStreamInternals, kStreamsPerPool>
 static std::array<LeakyStreamInternals, kStreamsPerPool>
     high_priority_streams[C10_COMPILE_TIME_MAX_GPUS];
 
+// External streams
+// We need some temporary LeakyStreamInternals to which to stage information
+// about external streams when the user attempts to set them as current.
+static thread_local LeakyStreamInternals external_streams[C10_COMPILE_TIME_MAX_GPUS];
```

Comment:
I don't understand. `current_streams` is a bunch of LeakyStreamInternals. Why can't you just store external streams there.

Reply:
`current_streams` looks to me like an array of _pointers_ to LeakyStreamInternals, but the actual LeakyStreamInternals objects must be stored somewhere else. Am I misunderstanding this?

Reply:
OK, this seems fair.

You don't have to fix this here, but it seems like current streams was misdesigned slightly. It seems to me that it would be better of us to store a Stream (instead of a pointer to LeakyStreamInternals) for current stream, then you don't have to create fake LeakyStreamInternals every time you set an external stream just to appease the internal data representation. The cost is that you have to now recompute the correct LeakyStreamInternals struct to get the true cuda stream on normal queries (I guess that's why it was done this way to begin with), but this should just be a small amount of bit twiddling and then some pointer arithmetic, so it ought not to matter as far as perf.

Reply:
Sure, how can I turn down a chance for a nice refactor? ;)

And, in fact, it gets even simpler: `current_streams` can just become an array of `StreamId`s! This way, everything falls nicely into place: no more "extracting" the stream id given the pointer, no need for a staging area for external streams, no need to store the device index in leaky streams, ...